File: SortArrayParity.py
# Given an integer array nums, move all the even integers at the beginning of the array followed by all the odd integers.


# Two Pass
# Intuition and Algorithm

# Write all the even elements first, then write all the odd elements.


class Solution(object):
    def sortArrayByParity(self, A):
        return ([x for x in A if x % 2 == 0] +
                [x for x in A if x % 2 == 1])



# In-Place
# Intuition

# If we want to do the sort in-place, we can use quicksort, a standard textbook algorithm.

class Solution(object):
    def sortArrayByParity(self, A):
        i, j = 0, len(A) - 1
        while i < j:
            if A[i] % 2 > A[j] % 2:
                A[i], A[j] = A[j], A[i]

            if A[i] % 2 == 0: i += 1
            if A[j] % 2 == 1: j -= 1

        return A

File: SortArrayParity.java
/**
Given an integer array nums, move all the even integers at the beginning of the array followed by all the odd integers.

Return any array that satisfies this condition.
*/

/**
my approach: create a new array and a variable to track its growthh
loop through the array, add even elements first to the new array
then add the odd elements
*/

class Solution {
    public int[] sortArrayByParity(int[] nums) {
        int[] newnums = new int[nums.length];
        int n = 0;
        for(int i = 0; i< nums.length; i++){
            if(nums[i]%2==0){
                newnums[n++] = nums[i];
            }
        }
        for(int i = 0; i< nums.length; i++){
            if(nums[i]%2==1){
                newnums[n++] = nums[i];
            }
        }
        return newnums;
        
    }
}
